pilot/pkg/model/sidecar_test.go

```diff
@@ -1465,6 +1466,82 @@ func TestCreateSidecarScope(t *testing.T) {
 	}
 }
 
+type ServiceOption struct {
+	Hostname  string
+	ports     []*Port
+	registry  provider.ID
+	namespace string
+}
+
+func makeService(args ServiceOption) *Service {
+	return &Service{
+		Hostname: host.Name(args.Hostname),
+		Ports:    args.ports,
+		Attributes: ServiceAttributes{
+			Namespace:       args.namespace,
+			ServiceRegistry: args.registry,
+		},
+	}
+}
+
+func TestAddService(t *testing.T) {
+	svc1 := makeService(ServiceOption{Hostname: "svc1"})
+	svc2 := makeService(ServiceOption{Hostname: "svc2"})
+	svc3 := makeService(ServiceOption{Hostname: "svc3", ports: port8000, registry: provider.External})
+	svc3DiffPort := makeService(ServiceOption{Hostname: "svc3", ports: port9000, registry: provider.External})
+	svc3DiffNs := makeService(ServiceOption{Hostname: "svc3", namespace: "test", ports: port8000, registry: provider.External})
+	svc3K8s := makeService(ServiceOption{Hostname: "svc3", namespace: "test", ports: port8000, registry: provider.Kubernetes})
+
+	twoPorts := append(port8000, port9000...)
+	svc3twoPorts := makeService(ServiceOption{Hostname: "svc3", ports: twoPorts, registry: provider.External})
+
+	testcases := []struct {
+		name     string
+		services []*Service
+		expected []*Service
+	}{
+		{
+			name:     "separate services",
+			services: []*Service{svc1, svc2},
+			expected: []*Service{svc1, svc2},
+		},
+		{
+			name:     "two services from SE with same host, same namespace, same port",
+			services: []*Service{svc3, svc3},
+			expected: []*Service{svc3},
+		},
+		{
+			name:     "two services from SE with same host, same namespace, different ports",
+			services: []*Service{svc3, svc3DiffPort},
+			expected: []*Service{svc3twoPorts},
```

Comment:
I have some concerns about just merging port. Consider a Service with port 8080. I define a SE for the same hostname:

```
apiVersion: networking.istio.io/v1beta1
kind: ServiceEntry
metadata:
  name: shell
spec:
  hosts:
  - shell.default.svc.cluster.local
  ports:
  - number: 12345
    name: cstum
    protocol: TCP
  - number: 8080
    name: exists
    protocol: HTTP
  location: MESH_INTERNAL
  resolution: STATIC
  endpoints:
  - address: 2.2.2.2
```

Now I get a cluster on ports 12345 and 8080. But the endpoints for 8080 are just the k8s pods, not `2.2.2.2`. This seems inconstent.

It seems like we should not merge at all. It is too tricky, with different resolution modes, different protocols on the same port, etc.

It seems like the original PR (https://github.com/istio/istio/pull/18111) was actuallying doing it for synthetic Sidecar services. So something like this:
```
  egress:
  - port:
      number: 9080
      protocol: HTTP
      name: egresshttp
    hosts:
    - "prod-us1/httpbin"
  - port:
      number: 12345
      protocol: HTTP
      name: egresshttp
    hosts:
    - "prod-us1/httpbin"
```

Not actually merging 'real' Services.

cc @ramaraochavali

Reply:
SE itself has already make istio very complex, as we allow creating SE with same hosts in any namespace with any ports.

Not to mention it can be used to define k8s svc as well.  Here I am not meant to solve the mess. I want to fix the inconsistent. 

Regarding to the mess, we should draft a higher design, and maybe need to break some compatibility if we want to make it clean.

Reply:
BTW, I have talked about the merging difficulty again and again with @zirain. And we haven't come up with a better solution.

Reply:
Totally agree, but I think in solving the inconsistency we may be expanding the wrong behavior? I think the merging was only intended to merge sidecar.egress synthetic imports , but now we fully extend it to all services which may not be great...

Reply:
Seems like there is no real use case for merging ports in real services?

Reply:
Agreed, if we want to remove merging ports we should have a way to warn users if they did use SE in this way.